lensless/mask.py

```diff
@@ -0,0 +1,321 @@
+            u_in=self.mask,
+            wv=self.wavelength,
+            d1=self.feature_size,
+            dz=self.distance_sensor,
+            dtype=np.float32
+        )
+        pass
+    
+
+    def shape(self):
+        """
+        Shape of the mask.
+        """
+        return self.mask.shape
+    
+
+    def phase_retrieval(self, lambd, d1, dz, n=1.5, n_iter=10):
+        """
+        Iterative phase retrieval algorithm from the PhlatCam article (https://ieeexplore.ieee.org/document/9076617)
+
+        Parameters
+        ----------
+        lambd: float
+            wavelength (m)
+        d1: float
+            sample period on the sensor i.e. pixel size (m)
+        dz: float
+            propagation distance between the mask and the sensor
+        n: float
+            refractive index of the mask substrate
+        n_iter: int
+            number of iterations
+        """
+        M_p = np.sqrt(self.psf)
+        for _ in progressbar.ProgressBar()(range(n_iter)):
+            M_phi = np.exp(1j * np.angle(fresnel_conv(M_p, lambd, d1, -dz, dtype=np.float32)[0]))
+            M_p = np.sqrt(self.psf) * np.exp(1j * fresnel_conv(M_phi, lambd, d1, dz, dtype=np.float32)[0])    
+        phi = np.angle(M_phi)
+
+        self.phase_mask = phi
+        self.height_map = lambd * phi / (2 * np.pi * (n-1))
+        pass
+
+
+
+
+
+class CodedAperture(Mask):
+    """
+    Coded aperture subclass of the Mask class
+    From the FlatCam article https://arxiv.org/abs/1509.00116
+    """
+
+    def __init__(self, 
+                 sensor_size_px: tuple, 
+                 sensor_size_m: tuple, 
+                 feature_size: tuple, 
+                 distance_sensor: float,
+                 wavelength: float, 
+                 method: str, 
+                 n_bits: int) -> None:
+        """
+        Coded aperture mask contructor (FlatCam).
+
+        Parameters
+        ----------
+        sensor_size_px: tuple (dim=2)
+            size of the sensor (px)
+        sensor_size_m: tuple (dim=2)
+            size of the sensor (m)
+        feature_size: float
+            size of the feature (m)
+        distance_sensor: float
+            distance between the mask and the sensor (m)
+        wavelength: float
+            wavelength (m)
+        method: str
+            pattern generation method (MURA or MLS)
+        n_bits: int
+            characteristic number for pattern generation
+            size = 4*n_bits + 1 for MURA
+                   2^n - 1 for MLS
+        """
+        
+        self.method = method
+        self.n_bits = n_bits
+
+        super().__init__(sensor_size_px, sensor_size_m, feature_size, distance_sensor, wavelength)
+
+
+    def create_mask(self):
+        """
+        Creating coded aperture mask using either the MURA of MLS method
+        """
+        assert self.method in ['MURA', 'MLS']
+        if self.method == 'MURA':
+            self.mask = self.squarepattern(4*self.n_bits+1)
```

Comment:
is this way of generating a mask produce a separable mask? i.e. a pattern that is the outer product of two 1D vectors

Reply:
I think so. At least it is exactly the same as MURA masks I see online.

Reply:
It's separable if we don't count the first row and first column